File: app/src/main/java/com/rodolfonavalon/canadatransit/model/database/converter/gson/DateTimeConverter.java
package com.rodolfonavalon.canadatransit.model.database.converter.gson;

import android.text.TextUtils;

import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonElement;
import com.google.gson.JsonParseException;
import com.google.gson.JsonPrimitive;
import com.google.gson.JsonSerializationContext;
import com.google.gson.JsonSerializer;

import org.joda.time.DateTime;
import org.joda.time.format.DateTimeFormatter;
import org.joda.time.format.ISODateTimeFormat;

import java.lang.reflect.Type;

/**
 * This is the Gson converter for the {@link DateTime} that serialize and deserialize from the
 * Realm database.
 */
public class DateTimeConverter implements JsonSerializer<DateTime>, JsonDeserializer<DateTime> {

    @Override
    public DateTime deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
        if (TextUtils.isEmpty(json.getAsString())) {
            return null;
        }
        final DateTimeFormatter formatter = ISODateTimeFormat.dateTimeParser();
        return formatter.parseDateTime(json.getAsString());
    }

    @Override
    public JsonElement serialize(DateTime src, Type typeOfSrc, JsonSerializationContext context) {
        final DateTimeFormatter formatter = ISODateTimeFormat.dateTimeParser();
        return new JsonPrimitive(formatter.print(src));
    }
}

File: app/src/main/kotlin/com/rodolfonavalon/canadatransit/model/database/converter/gson/DateTimeConverter.kt
package com.rodolfonavalon.canadatransit.model.database.converter.gson

import com.squareup.moshi.FromJson
import org.joda.time.DateTime
import com.squareup.moshi.Moshi
import com.squareup.moshi.ToJson
import org.joda.time.format.ISODateTimeFormat

/**
 * This is the [Moshi] converter for the [DateTime] that serialize and deserialize from the
 * Room database.
 */
class DateTimeAdapter {

    @FromJson fun fromJson(value: String): DateTime {
        val formatter = ISODateTimeFormat.dateTimeParser()
        return formatter.parseDateTime(value)
    }

    @ToJson fun toJson(dateTime: DateTime): String {
        val formatter = ISODateTimeFormat.dateTimeParser()
        return formatter.print(DateTime(dateTime))
    }
}
